src/Middleware/RequestThrottling/src/Internal/SemaphoreWrapper.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Internal
+{
+    internal class SemaphoreWrapper : IDisposable
+    {
+        private SemaphoreSlim _semaphore;
+        private int _waitingRequests;
+        private object _waitingRequestsLock = new object();
+
+        public SemaphoreWrapper(int queueLength)
+        {
+            _semaphore = new SemaphoreSlim(queueLength);
+        }
+
+        public async Task EnterQueue()
+        {
+            lock (_waitingRequestsLock)
+            {
+                _waitingRequests++;
+            }
+
+            await _semaphore.WaitAsync();
```

Comment:
We should save this task and check `IsCompletedSuccessfully` like we do in the middleware. If WaitAsync immediately completed successfully, you can save yourself from acquiring the lock twice to increment and then immediately decrement.

Reply:
I've made the change: if the semaphore immediately goes, the lock section is skipped.

Having duplicate structures immediately next to each other (Middleware -> RequestQueue) seems fishy. Theoretically I would like to move the logging into the RequestQueue and simplify the middleware logic, but for now it's probably more useful for logging to be at the higher level.